### PR TITLE
🐛 Include more device_owners when looking for a port for floating ip

### LINKED ISSUE #1985: Incorrect FloatingIP workflow
/kind bug

**What steps did you take and what happened:**
I tried capo build for `1d5d2d5e45462dab056e37a6c948361e81875ea9`. Some key details follow:

1) Created a `OpenStackFloatingIPPool` (non-relevant fields removed)

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
kind: OpenStackFloatingIPPool
metadata:
  name: osfipp
spec:
  floatingIPNetwork:
    id: c7c8509d-7083-41c9-b799-e30e855e9bc0
  reclaimPolicy: Delete
  # …
```

2) created a `MachineDeployment` and `OpenStackMachineTemplate`

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: OpenStackMachineTemplate
metadata:
  name: some
spec:
  template:
    spec:
      ports:
        - network:
            id: f16855bf-8ba1-4f75-ad8c-763e80134571
      floatingIPPoolRef:
        apiGroup: infrastructure.cluster.x-k8s.io/v1beta1
        kind: OpenStackFloatingIPPool
        name: osfipp
# …
```

✅ Floating IP was successfully created. Here we get correct data `fip.FloatingIP == "185.***.**.**", fip.FloatingNetworkID == "c7c8509d-7083-41c9-b799-e30e855e9bc0"`:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1d5d2d5e45462dab056e37a6c948361e81875ea9/controllers/openstackmachine_controller.go#L440-L443

❌ Here we get `port == nil` and an error "Failed while associating ip from pool: port for floating IP \"185.***.**.**\" on network c7c8509d-7083-41c9-b799-e30e855e9bc0 does not exist":
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1d5d2d5e45462dab056e37a6c948361e81875ea9/controllers/openstackmachine_controller.go#L450-L458

More details follow.

Here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1d5d2d5e45462dab056e37a6c948361e81875ea9/pkg/cloud/services/networking/port.go#L65

Openstack API returns the following (non-relevant fields skipped):

```json
{
  "ports": [
    {
      "device_id": "d1b99e45-991c-4143-93a3-9a8d3eddb416",
      "device_owner": "compute:nova",
      "fixed_ips": [
        {
          "ip_address": "10.21.10.29",
          "subnet_id": "616388c0-519f-418e-80b4-3687a546a65e"
        }
      ],
      "id": "0d1fe3bd-55f6-41d0-b879-a4071a15b5c0",
      "network_id": "f16855bf-8ba1-4f75-ad8c-763e80134571"
// …
    }
  ]
}
```

Please notice that we don't have a port associated with FIP network `c7c8509d-7083-41c9-b799-e30e855e9bc0`. And both FIP network ID and the FIP itself are not going to appear in the ports info because in our Openstack cloud floating IPs are not being added to ports directly. But NAT `185.***.**.**` → `10.21.10.29` would be set up.

If the new k8s node got FIP it could be found here:
`https://compute-api:8774/v2.1/TENANT_ID/servers/d1b99e45-991c-4143-93a3-9a8d3eddb416`

And the reply might be looking like this (non-relevant fields skipped):

```json
{ "server": {
    "id": "d1b99e45-991c-4143-93a3-9a8d3eddb416",
    "hci_info": {
      "network": [
        {
          "ips": [
            "10.21.10.29"
          ],
          "network": {
            "id": "f16855bf-8ba1-4f75-ad8c-763e80134571",
            "subnets": [
              {
                "ips": [
                  {
                    "address": "10.21.10.29",
                    "type": "fixed",
                    "version": 4,
                    "floating_ips": [
                      {
                        "address": "185.***.**.**",
                        "type": "floating",
                        "version": 4,
                      }
                    ]
                  } ] } ] } } ] } } }
```

Here it tries to find a fixed IP in the FIP network but in our openstack cloud all FIPs have `device_owner == "network:floatingip"` so it gets just an empty list:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/1d5d2d5e45462dab056e37a6c948361e81875ea9/pkg/cloud/services/networking/port.go#L71-L76

**What did you expect to happen:**
Successfully deployed k8s node with FIP attached.

**Anything else you would like to add:**
None so far. But please ask me any details. The issue is reproducible and I can add even more details if you want.

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): `1d5d2d5e45462dab056e37a6c948361e81875ea9`

- Cluster-API version: `1.6.3`
- OpenStack version: Virtuozzo (https://virtuozzo.com), based on Openstack Xena
- Minikube/KIND version: N/A
- Kubernetes version (use `kubectl version`): `1.29.3`
- OS (e.g. from `/etc/os-release`): Talos (https://talos.dev) `1.6.7`

### PR DESCRIPTION
**Pull request by bilbobrovall**
**What this PR does / why we need it**: Adds support in `GetPortForExternalNetwork` for more distributed/ha-routers

**Which issue(s) this PR fixes**: Fixes #1985 

I'm not 100% sure if we need all these device_owners or if we need some additional, I got them from https://github.com/openstack/neutron-lib/blob/master/neutron_lib/constants.py#L74 but didn't find any good documentation on them.

**Special notes for your reviewer**:
1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:
- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

File: pkg/cloud/services/networking/port.go [hunk 1]
```diff
@@ -58,6 +59,19 @@ func (s *Service) GetPortFromInstanceIP(instanceID string, ip string) ([]ports.P
 	return s.client.ListPort(portOpts)
 }
 
+type PortListOpts struct {
+	DeviceOwner []string `q:"device_owner"`
+	NetworkID   string   `q:"network_id"`
+}
```
Comment (dulek) [inline pkg/cloud/services/networking/port.go:65]: @pierreprinetti: I know you've been looking at this from Gophercloud perspective. Is this the way to go here?

**Comment by dulek** [comment]
Makes sense to me, tiny request for clarification in the comments.

/lgtm

File: pkg/cloud/services/networking/port.go (hunk 1 above)
Comment (mdbooth) [reply pkg/cloud/services/networking/port.go:65]: Eventually hopefully not, but I think this is the only way to go currently.

**Pull request merged**